rfcs/20210930-enable-onednn-ops.md

```diff
@@ -0,0 +1,194 @@
+
+## User Benefits
+
+Enabling oneDNN-based operations will bring benefits of TF-oneDNN to default TensorFlow packages.
+* **Performance speedups on x86 CPUs.** oneDNN’s primitives are highly optimized. Its just-in-time compilation and dynamic dispatch also let it use available hardware instructions efficiently. Enabling custom oneDNN ops achieves throughput speedups up to [3⨉ in inference and 2.4⨉ in training](https://medium.com/intel-analytics-software/leverage-intel-deep-learning-optimizations-in-tensorflow-129faa80ee07) on a Cascade Lake CPU.
+* **Timely support of new hardware features in Intel CPUs.** oneDNN can use new Intel hardware instructions as soon as they are launched. TF-oneDNN receives the same benefit.
+* **Mixed precision computation on CPU.** Default TF CPU ops don’t support bf16/fp32 mixed precision computations yet. oneDNN ops enable this feature on CPU. We observed additional [1.7-1.9⨉ speedups](https://blog.tensorflow.org/2020/06/accelerating-ai-performance-on-3rd-gen-processors-with-tensorflow-bfloat16.html) with bfloat16 hardware vectorization on a Cooper Lake CPU.
+* **Additional int8 / quantization support on CPU.** oneDNN ops have more int8 coverage than default TF CPU ops. We also observed [3-4⨉ speedups](https://www.intel.com/content/www/us/en/artificial-intelligence/posts/accelerating-tensorflow-inference-with-intel-deep-learning-boost-on-2nd-gen-intel-xeon-scalable-processors.html) on top of software quantization benefits with int8 hardware vectorization on a Cascade Lake CPU.
+
+## Design Proposal
+
+The design consists of three main components: oneDNN custom ops, graph rewrite passes, and a software-managed cache.
+
+**oneDNN custom ops** are TensorFlow ops that perform some set up (e.g., creating oneDNN compute primitives, memory descriptors, etc) before calling oneDNN routines.
+* **Visibility:** These custom ops are hidden from users in the Python layer, cannot be saved in SavedModels, and are only introduced into the TensorFlow graph by graph rewrite passes in the C++ layer.
+* **Scope and signature:** Most ops have a 1:1 mapping to a standard TensorFlow op, with the same op signatures (inputs, attributes, outputs, etc). The rest are fused operations (one op is equivalent to a sequence of standard TensorFlow ops).
+* **Expected tensor format:** oneDNN ops process tensors directly in TensorFlow’s native format (NHWC), e.g., no tensor layout conversions.
```

Comment:
I understand that for activation, we want to hide the blocked layout completely.  But for weight, I believe that framework may want to cache the reordered weights for the inference task.  Can you elaborate the expectation for weight layout?